### DISCUSSION TITLE
Use custom toolchain - without rustup

### DISCUSSION THREAD

**Discussion by otavio** [cross-rs/cross#1082] [Q&A]
Hello,

I am trying to use a custom toolchain with `cross` but couldn't find a way to do so. I get:

```console
% cross build --target i686-unknown-linux-gnu
Error:
   0: could not execute `rustup toolchain list`
   1: No such file or directory (os error 2)
```

Looking at the source, it seems something like this is used on `rustc` bisect, but I couldn't make it work. Any suggestion?

**Comment by Alexhuszagh** [comment]
Have you been using `CROSS_CUSTOM_TOOLCHAIN=1`? I believe this still might not work, but a minor modification should be able to fix this. The only use of `rustup` when using `cargo-bisect-rustc` or `CROSS_CUSTOM_TOOLCHAIN=1` should be in [lib.rs](https://github.com/cross-rs/cross/blob/ac7fbe80f71f9b4f6831a65825fed645fef1d875/src/lib.rs#L581-L588):

```rust
let installed_toolchains = rustup::installed_toolchains(msg_info)?;

if !installed_toolchains
    .into_iter()
    .any(|t| t == toolchain.to_string())
{
    rustup::install_toolchain(&toolchain, msg_info)?;
}
```

Toggling that line to only work if `is_custom` is not true should fix the issue. I think that's the last piece of code that relies on rustup with a custom toolchain. However, if possible, maybe add the toolchain to rustup itself? Rustup can handle custom toolchains itself, which can be very handy when building Rust from source:

```bash
$ rustup toolchain link --help
rustup-toolchain-link 
Create a custom toolchain by symlinking to a directory

USAGE:
    rustup toolchain link <toolchain> <path>
```

This way, you can still change Rust toolchains easily while also being able to use custom builds. This way, the linked toolchain would also be available in `rustup toolchain list`.

**Reply by Alexhuszagh** [reply]
There's actually another, separate issue. I'll see if I can provide a working solution when I get back. Basically, we need a rustup install for a lot of things, but if you use a custom toolchain, we likely also need to ensure it works without `rustup`. Currently, using the system Rust on Fedora 36, I get that `cross` expects the Rust toolchain to be found at `/stable-x86_64-unknown-linux-gnu`, which is clearly not where it is. `rustc --print sysroot` correctly reports that it is at `/usr`.

**Reply by otavio** [reply]
I found a similar thing. I sent https://github.com/cross-rs/cross/pull/1083 as draft with current change. See:

```console
cross build -v --target i686-unknown-linux-gnu
+ cargo metadata --format-version 1 --filter-platform i686-unknown-linux-gnu
+ rustc --print sysroot
+ /nix/store/c49n0m1hl7c1x1kpjqyhncnypbpspvzn-rust-mixed/bin/rustc
...
docker: Error response from daemon: error while creating mount source path '/nix/store/stable-x86_64-unknown-linux-gnu': mkdir /nix/store/stable-x86_64-unknown-linux-gnu: read-only file system.
+ rustup component list --toolchain c49n0m1hl7c1x1kpjqyhncnypbpspvzn-rust-mixed
Error:
   0: could not execute `rustup component list --toolchain c49n0m1hl7c1x1kpjqyhncnypbpspvzn-rust-mixed`
   1: No such file or directory (os error 2)
```

**Reply by Alexhuszagh** [reply]
We've actually got a lot of issues here, some of which might be less easy to fix. Also, some of them will not work out-of-the-box on other container engines very well like Podman outside of a Linux host (since it only makes the home directory available when making a machine).

This means that `QualifiedToolchain::default` for an install at `/usr/bin/rustc` will produce a toolchain name of `usr` and a sysroot of `/stable-x86_64-unknown-linux-gnu`, not exactly what's desired. This is because `QualifiedToolchain::new` does:

```rust
    pub fn new(channel: &str, date: &Option<String>, host: &ImagePlatform, sysroot: &Path) -> Self {
        let mut this = Self {
            channel: channel.to_owned(),
            date: date.clone(),
            host: host.clone(),
            is_custom: false,
            full: if let Some(date) = date {
                format!("{}-{}-{}", channel, date, host.target)
            } else {
                format!("{}-{}", channel, host.target)
            },
            sysroot: sysroot.to_owned(),
        };
        this.sysroot.set_file_name(&this.full);
        this
    }
```

This line, `this.sysroot.set_file_name(&this.full);`, sets the name to `stable-x86_64-unknown-linux-gnu`, which messes with the sysroot entirely. I should have a working solution soon.

**Reply by Emilgardis** [reply]
this has now been implemented in https://github.com/cross-rs/cross/pull/1085

**Comment by otavio** [comment]
0.2.5 broke this again. #1085 has fixed it, but it fails again using the new release. Using the original commit I use:
```
export NIX_STORE=/nix/store
export CROSS_CUSTOM_TOOLCHAIN=1                                                                                                                                                                                                                                                   
```
However, using 0.2.5 it fails with:
```
% cross build -vv --target x86_64-unknown-linux-musl
+ cargo metadata --format-version 1 --filter-platform x86_64-unknown-linux-musl
+ rustc --print sysroot
+ rustup toolchain list
Error:
   0: could not execute `rustup toolchain list`
   1: No such file or directory (os error 2)
```

**Reply by Emilgardis** [reply]
#817 and #1085 have not been added to 0.2, so this is expected. ~~cross 0.2 doesn't know about `CROSS_CUSTOM_TOOLCHAIN`~~

**Reply by Emilgardis** [reply]
ah! My bad, CROSS_CUSTOM_TOOLCHAIN was added in 0.2.2

https://github.com/cross-rs/cross/pull/803

The fix was not backported
